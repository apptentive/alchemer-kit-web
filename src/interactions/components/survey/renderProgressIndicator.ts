/**
 * Renders the step indicator at the bottom of the survey.
 *
 * @param {number} questionCount - The number of questions to track progress for
 * @returns {HTMLUListElement} - The DOM Node
 */
const renderProgressIndicator = (questionCount: number) => {
  const progressIndicatorWrapper = document.createElement('div');
  progressIndicatorWrapper.classList.add('apptentive-progress__wrapper');

  // Render progress bar if the number of questions is greater than 10
  if (questionCount > 10) {
    const progressBarElement = document.createElement('progress');
    progressBarElement.classList.add('apptentive-progress-bar');
    progressBarElement.max = 100;
    progressBarElement.value = 0;

    progressIndicatorWrapper.append(progressBarElement);
    return progressIndicatorWrapper;
  }

  // Otherwise render the step indicator with the number of questions
  const progressStepElement = document.createElement('ul');
  progressStepElement.classList.add('apptentive-step-indicator');

  // Create a page indicator based on how many total segments are configured for this survey
  for (let i = 0; i < questionCount; i++) {
    const indicatorElement = document.createElement('li');
    indicatorElement.classList.add('apptentive-step-indicator__item');

    if (i === 0) {
      indicatorElement.classList.add('apptentive-step-indicator__item--current');
    }

    progressStepElement.append(indicatorElement);
  }

  progressIndicatorWrapper.append(progressStepElement);
  return progressIndicatorWrapper;
};

export default renderProgressIndicator;
