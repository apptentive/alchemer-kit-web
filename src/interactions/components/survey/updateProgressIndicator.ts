/**
 * Updates the progress indicator element based on the current question index
 *
 * @param {number} stepCount - The total number of steps being tracked
 * @param {number} nextIndex - The next step index the indicator should update to
 */
const updateProgressIndicator = (stepCount: number, nextIndex = 0) => {
  // Update the value of the progress bar if the count is greater than 10
  if (stepCount > 10) {
    const progressBarElement = document.querySelector<HTMLProgressElement>('.apptentive-progress-bar');

    if (progressBarElement) {
      progressBarElement.value = (nextIndex / stepCount) * 100;
    }

    return;
  }

  // Otherwise update the step indicators
  const indicatorClass = 'apptentive-step-indicator__item';
  const currentModifier = 'current';
  const completeModifier = 'complete';

  // Remove the modifier class from the current page element
  const currentElement = document.querySelector(`.${indicatorClass}--${currentModifier}`);

  if (currentElement) {
    currentElement.classList.remove(`${indicatorClass}--${currentModifier}`);
  }

  const indicatorElements = document.querySelectorAll(`.${indicatorClass}`);

  if (indicatorElements.length > 0) {
    // Add the complete class to every indicator prior to the new current one
    // This is to make sure that when questions are skipped, they are not left in an "off" state
    Object.values(indicatorElements)
      .slice(0, nextIndex)
      .forEach((indicator) => {
        indicator.classList.toggle(`${indicatorClass}--${completeModifier}`, true);
      });

    indicatorElements[nextIndex].classList.add(`${indicatorClass}--${currentModifier}`);
  }
};

export default updateProgressIndicator;
