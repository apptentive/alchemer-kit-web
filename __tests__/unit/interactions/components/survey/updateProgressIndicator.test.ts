import renderProgressIndicator from '../../../../../src/interactions/components/survey/renderProgressIndicator';
import updateProgressIndicator from '../../../../../src/interactions/components/survey/updateProgressIndicator';

describe('updateProgressIndicator', () => {
  afterEach(() => {
    document.body.innerHTML = '';
  });

  test('properly updates step progress indicator', () => {
    document.body.append(renderProgressIndicator(5));
    updateProgressIndicator(5, 3);

    expect(document.body).toMatchSnapshot();
  });

  test('properly updates indeterminate progress indicator', () => {
    document.body.append(renderProgressIndicator(12));
    updateProgressIndicator(12, 3);

    expect(document.body).toMatchSnapshot();
  });

  test('properly handles a missing step progress indicator', () => {
    expect(() => updateProgressIndicator(5, 3)).not.toThrow();
  });

  test('properly handles a missing indeterminate progress indicator', () => {
    expect(() => updateProgressIndicator(12, 3)).not.toThrow();
  });

  test('properly handles a missing next index parameter for a step indicator', () => {
    expect(() => updateProgressIndicator(5)).not.toThrow();
  });

  test('properly handles a missing next index parameter for an indeterminate indicator', () => {
    expect(() => updateProgressIndicator(12)).not.toThrow();
  });
});
