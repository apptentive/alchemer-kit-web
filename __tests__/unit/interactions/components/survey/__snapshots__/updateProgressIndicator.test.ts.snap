// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateProgressIndicator properly updates indeterminate progress indicator 1`] = `
<body>
  <div
    class="apptentive-progress__wrapper"
  >
    <progress
      class="apptentive-progress-bar"
      max="100"
      value="25"
    />
  </div>
</body>
`;

exports[`updateProgressIndicator properly updates step progress indicator 1`] = `
<body>
  <div
    class="apptentive-progress__wrapper"
  >
    <ul
      class="apptentive-step-indicator"
    >
      <li
        class="apptentive-step-indicator__item apptentive-step-indicator__item--complete"
      />
      <li
        class="apptentive-step-indicator__item apptentive-step-indicator__item--complete"
      />
      <li
        class="apptentive-step-indicator__item apptentive-step-indicator__item--complete"
      />
      <li
        class="apptentive-step-indicator__item apptentive-step-indicator__item--current"
      />
      <li
        class="apptentive-step-indicator__item"
      />
    </ul>
  </div>
</body>
`;
